### ISSUE TITLE
Unexpected behavior when adding a song to the queue

### ISSUE THREAD

**Issue by Hellysonrp** [codetheweb/muse#403]
If you have a song originated from a playlist in the back of the queue (i.e. have already listened to it) and try to add any other song to the playlist, the song will be added in a position related to the position of the first song from a playlist in the queue. If the song from the playlist is the first song of the queue (of all songs, not just what is shown in the `queue` command), the new song will be added immediately after the current song; if it is the second on the queue, the new song will be added after the next song, and so on.
This behavior is exemplified in this ~poorly made~ GIF:

![add-song-behavior](https://user-images.githubusercontent.com/5438762/141933786-25a23cee-4d7f-422f-a15d-dbfc0c52826a.gif)

The cause of this behavior is this piece of code:

https://github.com/codetheweb/muse/blob/a4e67d63839154916783e7b4a5f6c65006e759f3/src/services/player.ts#L236-L253

In my own instance of the bot, I've replaced this piece of code with this:

```typescript
    } else {
      debug(`adding song - ${String(immediate)} - ${song.title}`);
      if (immediate) {
        const insertAt = this.queuePosition + 1;
        this.queue = [...this.queue.slice(0, insertAt), song, ...this.queue.slice(insertAt)];
      } else {
        this.queue.push(song);
      }
    }
```

I did not open a PR, however, because I don't know what the expected behavior is.
Why was that loop added in the first place?

**Comment by codetheweb** [comment]
Sorry, I'm having a bit of a hard time understanding what you're saying. If the below doesn't clear anything up, could you post a sequence of bot commands to demo the behavior you're describing?

-------

Going by the code snippets, I think you're asking why songs are added in the middle of a playlist in the queue rather than at the end of the queue like normal? 

That behavior is rather opinionated. The thinking was: if you've queued up a large playlist (and are currently in the middle of it) but want to introduce your friend to this cool new song unrelated to said playlist, adding it with `.p cool new track` would result in it playing next. After `[cool new track]` played, Muse would then continue on with the playlist.

I think this behavior is obsolete since the same thing can be accomplished using the play command along with the immediate flag (`.p [query] i`). I would be fine changing it to your code snippet.

**Comment by Hellysonrp** [comment]
An example (starting a queue with a playlist):

Consider a playlist with 3 songs:
* playlist song 1
* playlist song 2
* playlist song 3

With the queue empty, add the playlist to the bot:
`.p playlistUrl`

`.q` would return something like:
> **playlist song 1**
> playlist song 2
> playlist song 3

Let's say that we listened to the entire playlist. Let's add some other songs to the bot:

`.p some song 1`
`.p some song 2`
`.p some song 3`
`.p some song 4`

`.q` would return something like:
> **some song 1**
> some song 4
> some song 3
> some song 2

`some song 1` is not from the playlist, neither are the other ones.

---

Another example (playlist in the middle of the queue):

Consider the same playlist as before.

Let's start the queue with 5 songs:
`.p some song 1`
`.p some song 2`
`.p some song 3`
`.p some song 4`
`.p some song 5`

Let's add the playlist to the end of the queue:
`.p playlistUrl`

`.q` would return something like:
> **some song 1**
> some song 2
> some song 3
> some song 4
> some song 5
> playlist song 1
> playlist song 2
> playlist song 3

Let's say that we listened to the entire queue.
Let's add some more songs then:

`.p some song 6`
`.p some song 7`
`.p some song 8`
`.p some song 9`
`.p some song 10`
`.p some song 11`
`.p some song 12`
`.p some song 13`
`.p some song 14`
`.p some song 15`

`.q` would return something like:
> **some song 6**
> some song 7
> some song 8
> some song 9
> some song 10
> some song 11
> some song 15
> some song 14
> some song 13
> some song 12

---

`this.queue` inside `player.ts` is only clean if `.clear` is called.
`this.queue.some(...)` iterates through the whole `this.queue` array, from the beginning.
Apart from that, `insertAt` starts in the current position + 1, offsetting the new song based of something in the start of the queue and something in the middle of the queue at the same time.

This doesn't match with the intended behavior you described. It would be a bug if it weren't already obsolete.

> I would be fine changing it to your code snippet.

My snippet is a bit ugly the way it currently is.
I'll clean it up a bit and open a PR.
